internal/provisionertask/machine_fsm.go

```diff
@@ -0,0 +1,157 @@
+// Copyright 2025 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package provisionertask
+
+import (
+	"fmt"
+
+	"github.com/juju/collections/set"
+	"github.com/juju/errors"
+)
+
+// MachineState represents the state of a machine in the FSM.
+type MachineState string
+
+const (
+	// StatePending indicates the machine has been created but not yet started.
+	StatePending MachineState = "Pending"
+
+	// StateStarting indicates the machine is in the process of being started.
+	StateStarting MachineState = "Starting"
+
+	// StateCancellingStart indicates a request to stop the machine was received
+	// while it was still starting. The machine will transition to Stopping once
+	// the start operation completes.
+	StateCancellingStart MachineState = "CancellingStart"
+
+	// StateRunning indicates the machine has been successfully started and has
+	// an instance ID assigned.
+	StateRunning MachineState = "Running"
+
+	// StateStopping indicates the machine is in the process of being stopped.
+	StateStopping MachineState = "Stopping"
+
+	// StateDead indicates the machine has been stopped and removed. This is a
+	// terminal state.
+	StateDead MachineState = "Dead"
+
+	// StateFailed indicates the machine encountered an error during
+	// provisioning or operation. This is a terminal state.
+	StateFailed MachineState = "Failed"
+)
+
+// ErrIllegalTransition is returned when an illegal state transition is
+// attempted.
+var ErrIllegalTransition = errors.ConstError("illegal state transition")
+
+// allowedTransitions defines the valid state transitions in the FSM.
+// Each state maps to a set of states it can transition to.
+//
+// These transitions follow the following (mermaid) diagram:
+//
+// stateDiagram-v2
+//
+//	[*] --> Pending
+//	Pending --> Starting: machineChange(Alive & no InstanceId)
+//	Pending --> Dead: machineChange(Dead|Dying & no InstanceId)
+//	Starting --> Running: StartSuccess (instance created + info set)
+//	Starting --> Starting: StartTransientErr / backoff (retry loop internal to task today)
+//	Starting --> Failed: StartPermanentErr
+//	Starting --> CancellingStart: machineChange(Dead|Dying while in-flight)
+//	CancellingStart --> Stopping: StartSuccess (instance now exists → compensate)
+//	CancellingStart --> Dead: StartErr(any) (no instance to stop)
+//	Running --> Stopping: machineChange(Dead|Dying|Remove)
+//	Stopping --> Dead: StopSuccess (MarkForRemoval done)
+//	Stopping --> Stopping: StopTransientErr / backoff
+//	Stopping --> Failed: StopPermanentErr
+//	Dead --> [*]
+//	Failed --> [*]
+var allowedTransitions = map[MachineState]set.Strings{
+	StatePending: set.NewStrings(
+		string(StateStarting),
+		string(StateDead),
+	),
+	StateStarting: set.NewStrings(
+		string(StateRunning),
+		string(StateFailed),
+		string(StateCancellingStart),
+		string(StateStarting), // idempotent for transient errors.
```

Comment:
No, don't make the graph cyclic. Also, there is no point, your `TransitionTo` already handles the fact that if the current transition matches the new one, then it's ok.